tests/autorestart_test/test_container_autorestart.py

```diff
@@ -0,0 +1,354 @@
+        # and the testing will exit
+        if program_status == "RUNNING":
+            pytest.fail("Failed to stop program '{}' before test".format(program_name))
+
+    logger.info("Program '{}' in container '{}' was stopped successfully"
+                .format(program_name, container_name))
+
+
+def is_hiting_start_limit(duthost, container_name):
+    """
+    @summary: Determine whether the container can not be restarted is due to
+              start-limit-hit or not
+    """
+    service_status = duthost.shell("sudo systemctl status {}.service | grep 'Active'".format(container_name))
+    for line in service_status["stdout_lines"]:
+        if "start-limit-hit" in line:
+            return True
+
+    return False
+
+
+def verify_autorestart_with_critical_process(duthost, container_name, program_name,
+                                             program_status, program_pid):
+    """
+    @summary: Killing a critical process in a container to verify whether the container
+              can be stopped and then restarted correctly
+    """
+    if program_status == "RUNNING":
+        kill_process_by_pid(duthost, container_name, program_name, program_pid)
+    elif program_status in ["EXITED", "STOPPED", "STARTING"]:
+        pytest.fail("Program '{}' in container '{}' is in the {} state, expected 'RUNNING'"
+                    .format(program_name, container_name, program_status))
+    else:
+        pytest.fail("Failed to find program '{}' in container '{}'"
+                    .format(program_name, container_name))
+
+    logger.info("Waiting until container '{}' is stopped...".format(container_name))
+    pytest_assert(wait_until(CONTAINER_STOP_THRESHOLD_SECS,
+                  CONTAINER_CHECK_INTERVAL_SECS,
+                  check_container_state, duthost, container_name, False),
+                  "Failed to stop container '{}'".format(container_name))
+    logger.info("Container '{}' was stopped".format(container_name))
+
+    logger.info("Waiting until container '{}' is restarted...".format(container_name))
+    restarted = wait_until(CONTAINER_RESTART_THRESHOLD_SECS,
+                           CONTAINER_CHECK_INTERVAL_SECS,
+                           check_container_state, duthost, container_name, True)
+    if not restarted:
+        if is_hiting_start_limit(duthost, container_name):
+            logger.info("{} hits start limit and clear reset-failed flag".format(container_name))
+            duthost.shell("sudo systemctl reset-failed {}.service".format(container_name))
+            duthost.shell("sudo systemctl start {}.service".format(container_name))
+            pytest_assert(wait_until(CONTAINER_RESTART_THRESHOLD_SECS,
+                          CONTAINER_CHECK_INTERVAL_SECS,
+                          check_container_state, duthost, container_name, True),
+                          "Failed to restart container '{}' after reset-failed was cleared".format(container_name))
```

Comment:
Suggest breaking this into two statements for readability:
```
    restarted = wait_until(CONTAINER_RESTART_THRESHOLD_SECS,
                           CONTAINER_CHECK_INTERVAL_SECS,
                           check_container_state, duthost, container_name, True)

    pytest_assert(restarted, "Failed to restart container '{}' after reset-failed was cleared".format(container_name))
```

Reply:
Great suggestion! Will update.